monocle/gitlab/mergerequest.py

```diff
@@ -0,0 +1,531 @@
+        # for instance each time the Pull Request X is extracted by this function
+        # (because it has been updated) its computed id must remane the same.
+        # Then the objects will updated in the DB and will not concidered as a
+        # new object.
+        #
+        # Change and Event's attributes type must be follow (see: db.py)
+        #
+        objects: List[Union[Change, Event]] = []
+        url = "https://dummy.com/change/id"
+        dummy_change = Change(
+            _id="123",
+            _type="Change",
+            number=1,
+            change_id="group/dummyrepo/1",
+            title="A dummy change",
+            text="Body text",
+            url=url,
+            commit_count=1,
+            additions=10,
+            deletions=0,
+            changed_files_count=1,
+            changed_files=[File(additions=10, deletions=0, path="README.md")],
+            commits=[
+                Commit(
+                    sha="12345",
+                    author=create_ident(url, "John", []),
+                    committer=create_ident(url, "John", []),
+                    authored_at="2020-04-11T07:01:15Z",
+                    committed_at="2020-04-11T07:00:15Z",
+                    additions=10,
+                    deletions=0,
+                    title="A dummy commit",
+                )
+            ],
+            repository_prefix="group",
+            repository_fullname="group/dummyrepo",
+            repository_shortname="dummyrepo",
+            author=create_ident(url, "John", []),
+            committer=create_ident(url, "John", []),
+            merged_by=create_ident(url, "Zuul", []),
+            branch="dummy-feature",
+            target_branch="main",
+            created_at="2020-04-11T07:00:15Z",
+            merged_at="2020-04-12T07:00:15Z",
+            updated_at="2020-04-12T07:00:15Z",
+            closed_at="2020-04-12T07:00:15Z",
+            state="MERGED",
+            duration=3600,
+            mergeable=None,
+            labels=None,
+            assignees=None,
+            approval=None,
+            draft=None,
+            self_merged=False,
+        )
+        dummy_event = Event(
+            _id="cce_1",
+            _type="ChangeCreatedEvent",
+            created_at="2020-04-11T07:00:15Z",
+            author=create_ident(url, "John", []),
+            repository_prefix="group",
+            repository_fullname="group/dummyrepo",
+            repository_shortname="dummyrepo",
+            url="https://dummy.com/change/id",
+            branch="dummy-feature",
+            target_branch="main",
+            number=1,
+            change_id="group/dummyrepo/1",
+            on_author=create_ident(url, "John", []),
+            on_created_at="2020-04-11T07:00:15Z",
+            changed_files=[SimpleFile(path="README.md")],
+            approval=None,
+        )
+        objects.extend([dummy_change, dummy_event])
+        return objects
+
+
+if __name__ == "__main__":
```

Comment:
I think the first step is to implement, here, in that function, an action that will fetch a single MR via a graphql query. This will help to validate how the GitLab graphql endpoint must be queried (low level details: need a token ? is there quotas system ?), but also it will help to explore the Graphql objects available to query/request a MR details.

To have that first step validated I'd like to be able to run the __main__ and give a repo + a MR id as parameter and see the MR details displayed in the console.

Second step will be to have that function able to crawl over all MRs on a given repository but since a specific date (ideally the updatedAt date). Ideally this date constraint should be passed to the GitLab grapqhql API query.

By experience, writing a crawler is difficult, so I really recommend to move forward step by step. Each API Gerrit/GitHub have there own behavior with associated bugs, so I guess it will be the same for GitLab API.

Reply:
Thanks @morucci. I will go with your suggested plan.